packages/curation-market/clients/js/index.js

```diff
@@ -65,69 +125,300 @@ class BondAlreadyExists extends Error {
+      validLicenses[i].licenseId,
+      holderAddress
+    )
+    if(holdsLicense) {
+      return true
+    }
+  }
+  return false
 }
 
 async function getBalance (bondId, holderAddress) {
   init()
   await window.ethereum.enable()
   holderAddress = holderAddress ? holderAddress : await walletOrSigner.getAddress()
   const bondHash = ethers.utils.id(bondId)
-  return await autoBond.getBalance(bondHash, holderAddress)
+  return (await autoBond.getBalance(bondHash, holderAddress)).toNumber()
+}
+
+// replaces buy
+async function buyLicense(
+  contributionId,
+  submissionCb,
+  confirmationCb,
+  amount = "auto",
+  maxPriceBP = 100 // number of basis points max should be above price shown
+) {
+  init()
+  const bytes32Id = ethers.utils.formatBytes32String(contributionId)
+  const purchasePrice = (
+    await squadController.contributions(bytes32Id)
+  ).purchasePrice
+  let buyPrice = purchasePrice
+  const supply = await totalSupplyOf(contributionId)
+  if (amount === "auto") {
+    amount = linearCurveAmount(supply, purchasePrice)
+  } else {
+    buyPrice = await priceOf(contributionId, amount)
+    if (buyPrice.lt(purchasePrice)) {
+      throw new Error("amount too low to buy a valid license")
+    }
+  }
+  const maxPriceBuffer = buyPrice.mul(maxPriceBP).div(10000)
+  const maxPrice = buyPrice.add(maxPriceBuffer)
+  console.log(
+    "TEST squadController.buyLicense",
+    contributionId,
+    amount.toString(),
+    maxPrice.toString()
+  )
+  const approveTx = await reserveToken.approve(bondingCurveFactoryAddress, maxPrice)
+  await provider.waitForTransaction(approveTx.hash)
+  const tx = await squadController.buyLicense(
+    bytes32Id,
+    amount,
+    maxPrice,
+    {gasLimit: 4000000}
+  )
+  submissionCb(tx)
+  return handleConfirmationCallback(tx.hash, confirmationCb)
 }
 
-async function buy (units, bondId, options = {}) {
+async function buy (units, bondId, submissionCb, confirmationCb, options = {}) {
   init()
   const bondHash = ethers.utils.id(bondId)
   const fullOptions = Object.assign({}, defaults, options)
-  return await autoBond.buy(
+  const tx = await autoBond.buy(
     units,
     bondHash,
     fullOptions
   )
+  submissionCb(tx)
+  return handleConfirmationCallback(tx.hash, confirmationCb)
+}
+
+async function redeemLicense(licenseId, submissionCb, confirmationCb) {
+  init()
+  const tx = await tokenClaimCheck.redeem(licenseId)
+  submissionCb(tx)
+  return handleConfirmationCallback(tx.hash, confirmationCb)
+}
+
+async function sellTokens(
+  contributionId,
+  amount,
+  minPrice,
+  submissionCb,
+  confirmationCb
+) {
+  init()
+  const tx = await squadController.sellTokens(
+    ethers.utils.formatBytes32String(contributionId),
+    amount,
+    minPrice,
+  )
+  submissionCb(tx)
+  return handleConfirmationCallback(tx.hash, confirmationCb)
 }
 
-async function sell (units, bondId, options = {}) {
+// replaces sell
+async function redeemAndSell(
+  licenseId,
+  minPrice,
+  redeemSubmissionCb,
```

Comment:
Possibly use this 4 callback pattern for buyLicense too given that we have 2 transactions (the approve and the by) probably not needed for the experiment